### DISCUSSION TITLE
Prevent a timeout on SQLAlchemy

### DISCUSSION THREAD

**Discussion by metropt** [fastapi/fastapi#8841] [Questions]
I have an application based on fastapi(0.59.0) and SQLAlchemy(1.3.18). We use postgresql as database. The max connections the db accept is 80 and since I have 4 uvicorn workers I set DATABASE_POOL_SIZE to 13 and max_overflow=4 so SQLAlchemy could use 68 connections.

If I do around 68 simultaneous requests the application stall. When it works it takes 
around 2.5 seconds to finished all the 68 requests. I logged the requests and I can confirm that all the requests arrived at the middleware however I can't determine if it stall at get_db or at the endpoint.

I notice that If I do db.close() at the end of the endpoint I could run 500 simultaneous requests without any problem.

How can I debug what is happening? It's make no sense that 68 requests takes 2.5seconds but than on request waits 30seconds for a connection.

I use a middleware for auth:
```
# Middlewares
@app.middleware('http')
async def check_auth(request: Request, call_next):
    """Adds a middleware to all requests to check auth credentials.

    Args:
        request (fastapi.Request): The incoming request
        call_next (types.FunctionType): The next function to call

    Returns:
        The response returned from the call_next function
    """
    response: Response = None
    # Get access token from session cookie
    token = auth.decode_access_token(request.cookies.get(ACCESS_COOKIE))

    logger.info("Request HTTP: "+str(request.body()))

    # If it's invalid
    if token is None:
        response = await call_next(request)
```

It is an example endpoint:
```
@router.get('/logs/admin')
def get_logs(
    id: int = Query(None, ge=0),
    page: int = Query(0, ge=0),
    page_size: int = Query(DEFAULT_PAGE_SIZE, ge=1),
    timestamp: int = Query(None, ge=0),
    live: bool = Query(None),
    db: Session = Depends(get_db),
):

    if not live:
        info = rf_logs_crud.get_paged_tag_logs(db, id, timestamp, page, page_size)
        #db.close() #added to test
        return info_obj
```

It is the db session manager.
```
from os import getenv

from sqlalchemy import MetaData, create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker

from fastapi.logger import logger
import logging
logger = logging.getLogger('gunicorn.error')

DATABASE_URL = getenv('DATABASE_URL')
DATABASE_POOL_SIZE = int(getenv('DATABASE_POOL_SIZE'))

engine = create_engine(DATABASE_URL, pool_size=DATABASE_POOL_SIZE, max_overflow=4)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

Meta = MetaData(bind=engine)
Base = declarative_base(bind=engine, metadata=Meta)

# DB Dependency
def get_db():
    """Yields a db session.

    Yields:
        db (app.db.SessionLocal): A database session
    """
    db = SessionLocal()
    try:
        yield db
    except:
        db.rollback()
    finally:
        db.close()
```

**Comment by TheBeege** [comment]
For seeing timing, just add some logging statements at each point you're curious about. 

For example:
```Python
def get_db():
    """Yields a db session.

    Yields:
        db (app.db.SessionLocal): A database session
    """
    start_time = time.time()
    db = SessionLocal()
    logging.debug('[Dependency] Time to create session: %f', time.time() - start_time)
    try:
        yield db
    except:
        logging.debug('[Dependency] Time to hit exception: %f', time.time() - start_time)
        db.rollback()
    finally:
        logging.debug('[Dependency] Time to hit finish yield: %f', time.time() - start_time)
        db.close()
```

I recommend doing something similar in your middleware and path operation to help isolate the timing issue.

If you want to see timing per request, in your middleware, create some uuid and set it on `request.state.x` where `x` is some property you want to use to store the request ID. Include that in your logging, and you can group by individual requests.

Curious to hear about your findings, so do please post back later.

(Also, adding "Python" immediately after the backticks in your code snippet will do Python syntax highlighting. It makes it much easier to read. I strongly recommend 😄 )
